### ISSUE TITLE
Is it normal for games to crash if played on a read only filesystem?

### ISSUE THREAD

**Issue by i30817** [adventuregamestudio/ags#799]
I made a experiment and just squashed the whole ags game dir (all 60gb of it, into 37 gb). Anyway, Quest For Infamy crashes then on the 'new game' button pressed in this enviroment. 

So can game devs write to the game dir; or does the engine do it as a matter of course? The first can't be really fixed (except with a copy-on-write filesystem), but   the second may. The strange thing is that'd i'd swear AGS doesn't write any files on the game dir as a matter of course, but maybe there is some temporary file shenanigans. Anyway i'm building a debug version to see if there is a clue but the game crashed on script code, so probably not.

**Comment by i30817** [comment]
Mmh, the debug build crashes even earlier at startup. The release build waited until trying to start the game from the main menu, i guess the debug build does something more:
```

(gdb) r
Starting program: /media/i30817/Huggin/Documents/projects/ags/build/ags 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00005555555ac8a1 in Test_File ()
    at /media/i30817/Huggin/Documents/projects/ags/Engine/test/test_file.cpp:58
58	    out->WriteInt16(10);
(gdb) bt
#0  0x00005555555ac8a1 in Test_File ()
    at /media/i30817/Huggin/Documents/projects/ags/Engine/test/test_file.cpp:58
#1  0x00005555555ac826 in Test_DoAllTests ()
    at /media/i30817/Huggin/Documents/projects/ags/Engine/test/test_all.cpp:28
#2  0x000055555558984e in ags_entry_point (argc=1, argv=0x7fffffffdf58)
    at /media/i30817/Huggin/Documents/projects/ags/Engine/main/main.cpp:385
#3  0x000055555558843a in main (argc=1, argv=0x7fffffffdf58)
    at /media/i30817/Huggin/Documents/projects/ags/Engine/main/main_allegro.cpp:5
(gdb) 

```

**Comment by ivan-mogilko** [comment]
Are you running Debug build? It runs automatic tests that write some stuff. I made a ticket for removing these out to google tests earlier.

**Comment by i30817** [comment]
The debug build crashes earlier but the release build does too when transitioning from the main menu to in-game in this game. It may be specific to QFI. I'll try the code modification to see if i can get a good stacktrace of the earlier bug.

**Comment by sonneveld** [comment]
I recall that we test if the save game directory is writable. I think a tmp file is also involved in some cases.

**Comment by ivan-mogilko** [comment]
> So can game devs write to the game dir; or does the engine do it as a matter of course? The first can't be really fixed (except with a copy-on-write filesystem), but the second may. 

Here's the thing, a while ago I implemented path remapping that causes any game that tries to write to installation directory to write into "APPDATADIR" (ags own tag) instead.
However, there's a way for both developer and **player** to override it, by setting up "user_data_dir" and "shared_data_dir" in config file. This was added by request so that people could e.g. have game saves in same directory for their preferences, or **if they make steam version** that wants to have all files in one folder for syncing.

**Comment by ivan-mogilko** [comment]
Of course the game should not crash like that anyway.

**Comment by ivan-mogilko** [comment]
What kind of "crash" occurs on "New game" in QFI thought? Is it a game error message, or internal program error? That may be an additional bug here.

**Comment by i30817** [comment]
the stacktrace is unusuable because it comes from script, even with a debug build.

I'd forgotten about those properties... and now i think i'll have to recompress and 'sed' the acsetup.cfg(s)... :(


this is the crash i mentioned, in the debug build after disabling that test function.
```
ci_find_file: Looked for QFI.exe in rough /media/i30817/Ags/Quest For Infamy/, found diamond QFI.exe.
Lagging! Missed 8 ticks!
/media/i30817/Huggin/Documents/projects/ags/build/ags(+0x122947)[0x555555676947]
/media/i30817/Huggin/Documents/projects/ags/build/ags(+0x15d0d9)[0x5555556b10d9]
/media/i30817/Huggin/Documents/projects/ags/build/ags(+0x15d222)[0x5555556b1222]
/media/i30817/Huggin/Documents/projects/ags/build/ags(+0x15d561)[0x5555556b1561]
/media/i30817/Huggin/Documents/projects/ags/build/ags(+0x15d78e)[0x5555556b178e]
/media/i30817/Huggin/Documents/projects/ags/build/ags(+0x15dd6f)[0x5555556b1d6f]
/media/i30817/Huggin/Documents/projects/ags/build/ags(+0x157643)[0x5555556ab643]
/media/i30817/Huggin/Documents/projects/ags/build/ags(+0x3596c)[0x55555558996c]
/media/i30817/Huggin/Documents/projects/ags/build/ags(main+0x20)[0x55555558840a]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7ffff56afb97]

[Thread 0x7fffdbd20700 (LWP 27915) exited]
An error has occurred. Please contact the game author for support, as this is likely to be a scripting error and not a bug in AGS.
(ACI version 3.5.0.12)

in "System.asc", line 144
from "System.asc", line 696
from "room2.asc", line 138

Error: Error running function 'room_Load':
Error: Null pointer referenced


[Thread 0x7fffc22cd700 (LWP 27917) exited]
[Thread 0x7fffc2c0f700 (LWP 27916) exited]
[Thread 0x7ffff0e81700 (LWP 27913) exited]
***** ENGINE HAS SHUTDOWN
[Thread 0x7ffff7fb6680 (LWP 27909) exited]
[Inferior 1 (process 27909) exited with code 0133]
(gdb) bt
No stack.
(gdb) 

```

**Comment by ivan-mogilko** [comment]
> in "System.asc", line 144
from "System.asc", line 696
from "room2.asc", line 138

Error: Error running function 'room_Load':
Error: Null pointer referenced

Hmm, I'll give 95% chance they are not testing for File.Open return value.


> About the shared data dir, can i place patches.dat (for the MIDI driver) there? How about a common acsetup?

Engine, or rather Allegro looks for patches.dat only in predefined locations, like near executable.
acsetup may be placed in standard user folder, as explained in our OPTIONS.md (XDG_HOME etc, if you are on linux)

**Comment by i30817** [comment]
> Engine, or rather Allegro looks for patches.dat only in predefined locations, like near executable. I don't know if there are other places it may hook it up.
home is what i'm currently using but i wish there was a way to make it look in the 'same dir as the data package (which might be the windows executable in windows)'

> **Here's the thing, a while ago I implemented path remapping that causes any game that tries to write to installation directory to write into "APPDATADIR" (ags own tag) instead**.
However, there's a way for both developer and player to override it, by setting up "user_data_dir" and "shared_data_dir" in config file. This was added by request so that people could e.g. have game saves in same directory for their preferences, or if they make steam version that wants to have all files in one folder for syncing.

This is not happening though, most especially because i removed all the old acsetup.cfg(s). The game does have a extra 'PlayTime.dat' file in the dir, which looks like a likely culprit of the file they want to open, so i'll try to see if removing it and then recreating a minisquash and mounting it allow the game to progress further. 

If it doesn't work, It's a bit of a pain to figure out not all games are ready for true read only, but that's what copy-on-write is for i guess

**Comment by i30817** [comment]
Yep... `shared_data_dir` to a writable dir made it start up fine and create a brand new Playtime.dat there.

Can this have a fallback to the linux save if it's not writable?

**Comment by i30817** [comment]
I meanwhile found many games that do this by testing around my squash to see if they startup. Blackwell 1-4, King's quest remakes 1-2, Quest for Yrogl, some others. I suspect that many of them use 'redirectors' to display achievements in game without steam and needed something to write to.

**Comment by ivan-mogilko** [comment]
Regarding libagscreditz, since the sources are available probably this could be reimplemented as a portable code similarily to how JJS reimplemented flashlight and few other plugins.

**Comment by ivan-mogilko** [comment]
> I meanwhile found many games that do this by testing around my squash to see if they startup. Blackwell 1-4, King's quest remakes 1-2, Quest for Yrogl, some others. I suspect that many of them use 'redirectors' to display achievements in game without steam and needed something to write to.

Perhaps the solution for the engine could be to detect that destination is readonly and fallback to writable location (assuming there's a default one).

**Comment by i30817** [comment]
> Regarding libagscreditz, since the sources are available probably this could be reimplemented as a portable code similarily to how JJS reimplemented flashlight and few other plugins.

Libags fire is another one, though that is not available with source. However, the readme gives a list of all the functions in depth and what they do (click on the zip, and open the html) http://americangirlscouts.org/agsresources/AGS%20resources/plugins/

**Comment by i30817** [comment]
I was asked to post a clarification about a request on another linked thread here; i'll try to be clear.

Although it's possible to run all mentioned games that crash on a read-only filesystem by using `shared_data_dir` in the config file to point to some writable dir instead of the game dir\*, since that setting only accepts hardcoded paths, using it in that manner would mean that as soon as i changed the login user, or if i moved the filesystem to another computer the write could fail if the hardcoded dir was inside $USER (and thus read only on another login) or if it didn't exist at all (in another computer). Even if it created all the dirs on the 'doesn't exist case' it would be bad imo.

So i'm proposing a alternative; make `shared_data_dir` have a symbolic setting that is instantiated at runtime. For instance `shared_data_dir=game_save_dir` (or similar) would create the new file(s) the games tries to create on `/home/$USER/.local/share/ags/$GAME/` (on linux) and solve the issue once and for all if the user wants it.

I'm even considering requesting making this behavior the default (`shared_data_dir` being redirected to the game save dir), but i'm unsure if some games depend on it being set to the game dir by 'default'. I've seen games use it as a subdirectory, but never games that had extra data without setting `shared_data_dir` on the config, but maybe i never noticed them


\*since the games that crash crash on writing/creating the file, apparently all of them work fine if you don't have the file they need; they just initialize it on a base state. There maybe some that only read a existing/from the installer file from the 'shared_data_dir' though.

**Comment by morganwillcock** [comment]
Isn't this just an issue that someone is setting shared_data_dir, which is meant for use only on certain distribution platforms, and then releasing it outside of the those platforms?

There are already redirections in place to stop writes to the game folder, because the same problem exists on Windows (where the installation directory may not be writeable by the user running the game).

The problem you are describing sounds more like packaging error by the author, where they choose to override the engines choices for data paths in favour of their own. Maybe it is just that documentation should be clearer.

**Comment by i30817** [comment]
I'm looking to make my games portable across computers (and hopefully OSes) ofc. Part of it is being able to adapt the runtime paths to always have a writable dir. If at the same time i want to apply filesystem compression, i'm screwed if there is not a symbolic path handle to place on `shared_data_dir`.

**Comment by i30817** [comment]
All of my examples above (Quest For Infamy, Blackwell 1-4, King's quest remakes 1-2, Quest for Yrogl) don't define `shared_data_dir` on the original .cfgs (well, i edited them and forgot about them but I know they didn't have it because all of their engine data is on a single dir).

I want to remind you, in case it's confusing, that this bug only occurs if I, the user, puts the game on a read only dir, therefore it's my responsibility to change the `shared_data_dir`. I just can't do it portably without a runtime path resolution symbolic option to put on the right hand side of the `=`.

ivan said this up above:

> Here's the thing, a while ago I implemented path remapping that causes any game that tries to write to installation directory to write into "APPDATADIR" (ags own tag) instead.
> However, there's a way for both developer and **player** to override it, by setting up "user_data_dir" and "shared_data_dir" in config file. This was added by request so that people could e.g. have game saves in same directory for their preferences, or **if they make steam version** that wants to have all files in one folder for syncing.

In my experience, the `APPDATADIR` mentioned here *is* the game directory; but maybe only because i'm starting the game by using the cmd line to `cd` to the game dir and calling the (system installed) `ags` - maybe it's supposed to be the dir where 'ags' is run from, or where the executable is. Even then, i'd want to change it to the symbolic option that points to the runtime save dir, because those dirs can easily be random or readonly (in fact the second nearly always is in linux).

**Comment by morganwillcock** [comment]
> This is a bit confusing to parse but it means that APPDATADIR is usually the game dir, but can be replaced by setting shared_data_dir.

I believe he was referring to the tokens that are expanded as part of file operations:

`$APPDATADIR$` expands to a system writeable location
`$INSTALLDIR$` expands to the game directory
`$SAVEGAMEDIR$` expands to the save directory

I'm not sure when these tokens appeared in relation to the script API, but if you try to write to relative paths this should already by redirected (I think this tries to solve the issue you describe). From the manual:
> An attempt to write file into relative path without specifying any location tag will make AGS to automatically remap such path into $APPDATADIR$

So perhaps there are two possibilties:
 - write behaviour also changes with requested script API level (so the redirection described is not in-effect)
- the write is explicitly referencing the installation directory (which will fail)

**Comment by i30817** [comment]
I think you're seeing a bug where there isn't one tbh. Isn't it 'normal' that `$APPDATADIR$` is the same as `$INSTALLDIR$` in the absence of a setting? If so, the behaviour is completely expected, and that was what i was assuming here.

**Comment by ivan-mogilko** [comment]
> This is a bit confusing to parse but it means that APPDATADIR is usually the game dir, but can be replaced by setting shared_data_dir.
>  Isn't it 'normal' that $APPDATADIR$ is the same as $INSTALLDIR$ in the absence of a shared_data_dir setting?

By default $APPDATADIR$ must point to system-specific common writeable location. This is what it does on Windows at least, and Windows version may be relied as an example in this regard.
This may not be a case for other ports though, where we could forgot to adjust this path default, or maybe did not know what would be best.

**Comment by i30817** [comment]
Well then, that's broken in linux then. As said, it always crashes if the game dir is read only which indicates it's set to the game dir and fallbacks to the save dir do not exist.

**Comment by i30817** [comment]
So what's $APPDATADIR$ supposed to be in windows anyway if not the game dir?

Because it's probably very annoying to have to setup explicit subdirs to prevent clashing same named files, so the 'shared directory' you mention is hopefully not really shared to the point it's the same between different games...

**Comment by morganwillcock** [comment]
> So what's $APPDATADIR$ supposed to be in windows anyway if not the game dir or the save dir?

It is shared, as in, multi-user writeable per game. It is equivalent in concept to %PROGRAMDATA% on Windows.

> As said, it always crashes if the game dir is read only which indicates it's set to the game dir and fallbacks (or a default) to the save dir do not exist, which is the obvious way to deal with it.

> Anyway, i'll leave it to you to decide what's best here. Default that errors out with message if read only forcing the user to use shared_data_dir, or fallback to the save dir.

I don't think anything is redirected to the save directory. Redirections go to $APPDATA$, because the redirections need to be writeable and multi-user (to mimic the result that the game directory was writeable). It might just be that the default path for $APPDATA$ on Linux isn't suitable and needs to be changed. Or maybe some games are specifically trying to write using the token $INSTALLDIR$ which, as documented, will fail.

**Comment by i30817** [comment]
>I don't think anything is redirected to the save directory. Redirections go to $APPDATA$, because the redirections need to be writeable and **multi-user** 
Ok, if you mean here 'this dir is different per user' this makes sense. If you mean 'this dir is the same for all users' it makes no sense, because a game writing a file, typically wants that file because of user interaction, for instance the AGDI games write a file of the 'points', so have 'achievements' etc. So starting a new game as a user and getting the points or achievements of the other user is a bit questionable to me. Not to mention that if that dir doesn't get a per-game subdir, there might be two games on a series (say the blackwell series) writing and reading to the same dir with the _same name_.

I'll assume you meant 'that dir is per user and per game like the save dir' for my sanity here.


> Or maybe some games are specifically trying to write using the token $INSTALLDIR$ which, as documented, will fail.

I don't think it's this because as mentioned setting `shared_data_dir` worksaround this and that setting is a $APPDATA$ override.

**Comment by morganwillcock** [comment]
Multi-user and writeable:
https://adventuregamestudio.github.io/ags-manual/File.html

**Comment by i30817** [comment]
Well... that's a thing. I'm going to reiterate the request to be able to set `shared_data_dir` to `user_data_dir` as a cfg interface to set $APPDATADIR$ to $SAVEGAMEDIR$. At least then i know games wont overwrite or accidentally read eachothers files...

**Comment by ivan-mogilko** [comment]
I guess both requests make sense and could be done altogether.
My concern is that linux etc users who got used to games write into game's dir may become confused or annoyed by this change. OTOH, there's a global config where one may add overriding `shared_data_dir = .` to force all games write into installdir, so workaround is present.

**Comment by i30817** [comment]
I'm being a bit uncharitable. May *most* games that write (there are very few ones) write to a subdir already because the devs thought of this, and maybe some devs are using $APPDATADIR$ shared responsibly (for instance imagine QFG remakes importing characters final saves automatically). 

Just because i've never seen a ags game using such tricks yet doesn't mean one doesn't exist now or in the future, so i'm likely to try the 'new default' to see what happens on the game i have verified that write.

**Comment by morganwillcock** [comment]
BTW, when you say squash are you using SquashFS?

**Comment by i30817** [comment]
Yeah.

**Comment by messengerbag** [comment]
> So what's $APPDATADIR$ supposed to be in windows anyway if not the game dir or the save dir?
> Because it's probably very annoying to have to setup explicit subdirs to prevent clashing same named files, so the 'shared directory' you mention is hopefully not really shared to the point it's the same between different games...

I think you got confused by morganwillcock's response here. On Windows, the default `%APPDATA%` path (possibly depending on certain Windows variables) is `C:/ProgramData/Adventure Game Studio/game_name`. In other words, it's not shared between different games (as long as they have different names).

**Comment by morganwillcock** [comment]
Well, this is where the naming issues comes in. %APPDATA% on Windows goes to your user profile, so this isn't related to the engine token $APPDATA$, which goes to a sub-folder of %PROGRAMDATA%.

<pre>
echo %PROGRAMDATA%
C:\ProgramData

echo %APPDATA%
C:\Users\mwillcock\AppData\Roaming
</pre>

There are also built-in redirections in Windows that will redirect writes to restricted areas into your user profile. So you can think you are writing to somewhere in %PROGRAMFILES% but you won't be, and no-one else will be seeing these changes.

**Comment by i30817** [comment]
Ok, so it doesn't have to be 'shared'. 

On linux the saves go to $HOME/.local/share/ags/$GAMENAME, which is private.

For the APPDIR i suggest $HOME/**.cache**/share/ags/$GAMENAME

also i suggest you use the [xdg config dirs standard](https://wiki.archlinux.org/index.php/$XDG_CONFIG_DIRS) if possible because users or distros sometimes want to move those dirs. It's actually easy to use without dependencies, just check if the $VARIABLE is set on the enviroment and if not default to the default location. If the user changes it, it's set on the env early on boot.

**Comment by ivan-mogilko** [comment]
> just check if the $XDG_DATA_HOME is set on the environment and if not set or empty default to the default location. 

AFAIK this is what we do for savegame locations right now.

**Comment by ericoporto** [comment]
Here -> https://github.com/adventuregamestudio/ags/blob/65c2a44681992fdca5ee31e0ed07b6269b057c29/Engine/platform/linux/acpllnx.cpp#L81

**Comment by i30817** [comment]
Right, got confused why i couldn't reproduce the crash i had before yesterday and i found what files AGDI in KQ2 remake is modifying and why you might not notice the crash. 

There is a file called 'intro.dat' in the dir. Inside it had a 'done' string. There is also a file named 'sk.dat' and inside it had 'not done'.

So this appears to be a 'can you skip the cutscene' hack, and when i was testing with a 'writable' dir it succeed in the write and no longer tried. Copying sk.dat into a new 'intro.dat' and making the file (not the dir, because making the dir 'access files' permission still succeeds at modifying the file) readonly 'restored' the crash.

In a squash partition ofc, you can't modify files at all.

This is a bit irrelevant to the bug itself, just something to be aware if you're testing this.

**Comment by i30817** [comment]
Ok this is weird. AGSLinux::GetAppOutputDirectory is called and returns the correct dir but AGSLinux::GetAppOutputDirectory is not called (but would return the correct dir).
```

Reading symbols from /media/i30817/Huginn/Documents/projects/ags/build/ags...
(gdb) break GetAppOutputDirectory
Breakpoint 1 at 0x3d1d2: GetAppOutputDirectory. (2 locations)
(gdb) r

```

**Comment by i30817** [comment]
It gets to here:
https://github.com/adventuregamestudio/ags/blob/3a7e17dc7239f075b7b10d7e4d697a73bfae9ca5/Engine/ac/file.cpp#L315

fails the else if and goes directly to the else where it fails again. Not sure how to read that code.

**Comment by ivan-mogilko** [comment]
Function name may be confusing, and I may change it. This function does not do what you may think:
<pre>
// Get default directory for program output (logs)
virtual const char *GetAppOutputDirectory() { return "."; }
</pre>

**Comment by i30817** [comment]
A override?

**Comment by ivan-mogilko** [comment]
GetAppOutputDirectory returns path to logs, not $APPDATADIR$

**Comment by ivan-mogilko** [comment]
> It gets to here:
> https://github.com/adventuregamestudio/ags/blob/3a7e17dc7239f075b7b10d7e4d697a73bfae9ca5/Engine/ac/file.cpp#L315
> 
> fails the else if and goes directly to the else where it fails again. Not sure how to read that code. The GameDataDirToken does appear to be "$APPDATADIR$" thou.

This code tests if script filename contains the token. If this condition fails, this means that in script they did not specify  $APPDATADIR$ at all.

**Comment by i30817** [comment]
Ok, then it's the else. From what i understand of that code, with `shared_data_dir` the `alt_path` is not created (the `if (read_only)` fails). Without, it attempts and succeeds to create a path 'to read' a file if not found on the default game App data dir.

The weird thing is that this `alt_path` is either not being used only for reading or the `parent_dir` above that if is also being set to the current game dir instead of what `parent_dir = MakeAppDataPath();` would indicate.

How do i print your engine version of String on gdb?

**Comment by i30817** [comment]
Ah. virtual const char *GetAllUsersDataDirectory() { return "."; }

This doesn't sound very correct?

**Comment by ivan-mogilko** [comment]
> Ah. `virtual const char *GetAllUsersDataDirectory() { return "."; }`
> 
> This doesn't look very correct?

This is correct, because it's base class, and it's the only default we may make for unknown system.
It should be overriden in Linux platform driver appropriately.

**Comment by i30817** [comment]
Ok, that's what's wrong then. Probably will changing that will fix everything, cross fingers.

**Comment by i30817** [comment]
BTW, the function inn Engine/test/test_all.cpp also needs to be disabled, because it's doing logs unconditionally in the game dir on debug builds.

That done, tests from the squash filesystem and a empty ags save dir:

KQ2 : worked, read the `intro.dat` file and wrote the altered one to the save dir with the game dir subdir
Quest for Infamy: worked, same with the `PlayTime.dat` file.
Quest for Yrolg: worked, same with the `muggit.dat` file.
Blackwell 1: worked, same with the `prog.bwl` file (this is probably a settings file)
Blackwell 2: worked, same with the `prog.bwu` file
Blackwell 3: worked, same with the `prog.bwc` file
Blackwell 4: worked, same with the `prog.bwd` file
Blackwell 5: worked, same with the `prog.bwe` and `eend.bwe` files
Primordia: worked, created a new file `prog.pri` on the save dir with the game dir subdir

Seems all correct...

**Comment by i30817** [comment]
BTW, i find the idea of `GetAllUsersDataDirectory` extremely weird. 

I *think* the reason that windows and osx got it and linux and android and ios didn't, is that the implementator tested on windows and osx, and found it 'necessary' (maybe because of game directory windows write protection).

but maybe on those other platforms assumed from the name 'oh this platform doesn't *have* shared dirs that are writable, always default to game dir'.

That aside, and now having proved the need for it, i don't understand why this is even necessary. Why have a 'All users data directory' when the savegame dir exists and is always guaranteed writable? Why even share game data across users when it's for writing?

**Comment by ivan-mogilko** [comment]
I am afraid more confusion is going on here.

First of all, it's not an idea of `GetAllUsersDataDirectory` at all. It's idea of $APPDATADIR$ path, that points to "shared data location".

The reason why `GetAllUsersDataDirectory` does not have its own implementation on Linux is probably trivial: after linux port was created nobody bothered setting it up correctly, and later nobody complained.

>  i don't understand why this concept is even necessary. Why have a 'All users data directory' function and make it different on windows and osx when the savegame dir exists and is always guaranteed writable and the cfg is already duplicated there (in linux at least)? Why even share game data across users when it's for writing?

It's a special case when game shares data among different players. Think of highscores, or level editor.

**Comment by i30817** [comment]
Then game devs are obviously using it wrong, because all the examples i found across 320 games were things that had no business being shared (but had bussiness having a write-redirection in case the target game dir was read only).

In short, just maybe is the engine using the $APPDATADIR$ path when the game dev just tried to write to the game dir and failed? That is *not* the same as trying to use a shared data dir; so the write redirection shouldn't go to a 'shared dir'. IMO.

**Comment by ivan-mogilko** [comment]
> That is not the same as trying to use a shared data dir; so the write redirection shouldn't go to a 'shared dir', but the save dir.

Maybe.

**Comment by ivan-mogilko** [comment]
> That is not the same as trying to use a shared data dir; so the write redirection shouldn't go to a 'shared dir', but the save dir

Maybe. It's not easy to define.
The remapping is only done for old games when the rule "writing into game dir is forbidden" was not present. So developers obviously cannot decide where it is going to be remapped.
Furthermore, some of these games were made before $APPDATADIR$ existed.

Unfortunately, engine cannot deduce whether this game data is supposed to be shared between users or not. Some data may be, while other not. For instance, one of the games I had in mind (named "Colorwise") had a level editor. My concern was that if I remap level editor files to savedir, then other users on same computer won't be able to see levels made by others.

Maybe it's not a big deal and it was a mistake taking this into account. In general, I think whole file sharing between users is not very well thought in AGS.

**Comment by i30817** [comment]
> Remapping to savedir will ofc work, but $APPDATADIR$ should be still fixed on Linux since some may have it used for "correct" purposes.

Yes, i agree (though maybe not for android and ios, since phones typically don't have multiple users). 

About the else case: i think the parent dir should be changed `parent_dir = get_save_game_directory();`.

Let me show my logic: there are more cases of devs writing to the game dir that get 'logic broken' that 1 game that *maybe* intended to share maps. For instance a new user in windows starts 'Quest for Infamy and finds that he already 'played' several hours in spite of never having played. Bug. A new user tries to pllay Kings Quest 2 remake and never sees the intro because another did. Bug. Etc.

**Comment by ivan-mogilko** [comment]
> Let me show my logic: there are more cases of devs writing to the game dir that get 'logic broken' that 1 game that maybe intended to share maps. For instance a new user in windows starts 'Quest for Infamy and finds that he already 'played' several hours in spite of never having played. Bug. A new user tries to pllay Kings Quest 2 remake and never sees the intro because another did. Bug. Etc.

Bugs like that are rather original game bugs. In fact, I could argue that by remapping into shared data directory we are keeping old behavior (even though it was logically faulty).

OTOH, if I remember correctly, in the past I was thinking, and suggesting, to add another config option - for these particular purposes - that tells engine where to remap old games to (savedir, or shared data dir).

**Comment by i30817** [comment]
I tested those i listed on the last post, (after removing the linux override for `GetAllUsersDataDirectory` I did and modifying `parent_dir = get_save_game_directory();` and they all worked instead of crashing.

So i'm pretty certain both things should be done. The linux version should get a `GetAllUsersDataDirectory` so the token will work (but that will be deceptive because *no* linux dir is shared and writable except /tmp, so it should be 'silently disabled' by making to the save dir).

and the else case should use `parent_dir = get_save_game_directory();`  because of the rationale above where there are many more bugs because of it than sacrifices (that don't crash) if the 'shared data dir' is used.

**Comment by ivan-mogilko** [comment]
> So i'm pretty certain both things should be done. The linux version should get a GetAllUsersDataDirectory so the token will work (but that will be deceptive because no linux dir is shared and writable except /tmp, so it should be 'silently disabled' by making to the save dir).

That's a problem I was never aware of, and IMO needs its own ticket to figure it out. If any game developer would ever expect to use $APPDATADIR$ for the purpose of sharing game data between computer users, they will get unexpected results on Linux. Maybe it was not best idea of a path token in the first place and may be redesigned in later versions.

**Comment by i30817** [comment]
I think a warning on the linux version when the path is retrieved is fine no? Or would that spam?

**Comment by ivan-mogilko** [comment]
> I think a warning on the linux version when the path is retrieved is fine no? Or would that spam?

That will be too late, they may have already designed the game with this functionality in mind.

**Comment by i30817** [comment]
There really isn't a practical solution (because having a shared writable dir would be a security issue by itself) but it should still work so people using the token, or linux users playing windows games (ie: me) don't get surprised by a game that crashes on a read-only dir that it defaulted to.

**Comment by i30817** [comment]
Speaking of multiuser security issues and the 'only writable and shared dir' this is a interesting read: http://www.chiark.greenend.org.uk/~peterb/uxsup/project/tmp-per-user/

So it's basically security kryptonite.

**Comment by messengerbag** [comment]
> About the else case: i think the parent dir should be changed parent_dir = get_save_game_directory();.

I don't know about any of this multi-user shared directory stuff, but I'd say there are good reasons to keep $APPDATADIR$ separate from the save game directory by default. Depending on how games use it, they can write custom configuration files, screenshots, generated outputs… all sorts of things that aren't in any way "savegames". Most importantly, they may be relying on reading data from this directory as well, so that deleting the files could have unintended consequences.

In my opinion, the expectation should be that you can delete the save game dir and only lose your saves, and that other files generated by the game are stored elsewhere.

**Comment by ivan-mogilko** [comment]
> > About the else case: i think the parent dir should be changed parent_dir = get_save_game_directory();.
> 
> I don't know about any of this multi-user shared directory stuff, but I'd say there are good reasons to keep $APPDATADIR$ separate from the save game directory by default.

I should clarify again, that this note above is referring to backwards-compatibility hack that remaps paths in installdir. This is used only for games made before AGS 3.3.5.

**Comment by i30817** [comment]
> In my opinion, the expectation should be that you can delete the save game dir and only lose your saves, and that other files generated by the game are stored elsewhere.

The 'custom configuration files' already go to the save dir in linux, because again, there is no 'shared custom dir'. Also this is *only* a backwards compatibility measure, so that games *not* using the token, ie: where you don't know what the author wanted, default to the save dir, which is much more sane because there are a bunch of them and 'sharing' data written like th8is causes logic bugs!

**Comment by ivan-mogilko** [comment]
> The 'custom configuration files' already go to the save dir in linux, because again, there is no 'shared custom dir'.

Engine writes user config to save dir, on all systems.
@messengerbag means custom config, written by game.

**Comment by morganwillcock** [comment]
> > About the else case: i think the parent dir should be changed parent_dir = get_save_game_directory();.
> 
> I don't know about any of this multi-user shared directory stuff, but I'd say there are good reasons to keep $APPDATADIR$ separate from the save game directory by default. Depending on how games use it, they can write custom configuration files, screenshots, generated outputs… all sorts of things that aren't in any way "savegames". Most importantly, they may be relying on reading data from this directory as well, so that deleting the files could have unintended consequences.
> 
> In my opinion, the expectation should be that you can delete the save game dir and only lose your saves, and that other files generated by the game are stored elsewhere.

Unfortunately the data is already mixed, even on Windows data that isn't saved games is going into the saved games folder. So really the saved games token is misnamed and going to the wrong location, and per user data is actually implemented here.  See #520.

**Comment by ivan-mogilko** [comment]
I think there's too much confusion going on about stuff in this thread... we better wrap this up.

I guess the paths could (should) be completely redesigned in future version(s) with better portability, possible needs and correctness in mind.

For now though we need at least one trivial fix to make Linux port at least comply to convention that games may be run from read-only location:

1. Fix Linux port by making $APPDATADIR$ point to actual safe writeable location.
After learning that Linux does not *technically* provide shared user dir, I see two options: either savedir, or extra dir in the same user's private space, located nearby. @i30817 said there's seem to be no point in latter; although maybe the point is at least in making them distinct in case game devs want them to be distinct?

Additionally, there's following issue:

2. For backwards compatibility, - that is if you run games made before ags 3.3.5 with the current engine, - there's an automatic remap that changes writing to game dir into writing to $APPDATADIR$. Question is, whether this is correct or these files should go to $SAVEGAMEDIR$.
Note that here were have mostly *logical* choice, because as we see on most platforms there's practically no difference (I think only Windows will have a significant difference - that $APPDATADIR$ is shared between users).

**Comment by i30817** [comment]
But again, it's a significant difference that fixes bugs, because it's a fallback that is assuming that a unusual situation (the dev wants to share created data) is *normal* instead of the exception. Which i hope my little survey of 321 games showed is false.

**Comment by ericoporto** [comment]
Erh, any chance of making a function to resolve a path using these special tokens? I needed `$SAVEGAMEDIR$` but file has no function like... `String File.GetPath(String pathWithTokens)`  that expanded the tokens into actual path. I can open a ticket for this.

**Comment by ivan-mogilko** [comment]
I opened 2 relevant prs:
#900,
#901

**Comment by ivan-mogilko** [comment]
May we consider this resolved for the time being, after #900 was merged?

**Comment by i30817** [comment]
Sure.
